src/Shape.cpp

```diff
@@ -22,6 +22,71 @@ bool Shape::intersect(const Ray &, float t0, float t1, double& distance) const {
     return false;
 }
 
+bool Shape::intersect(const Vector3& p) const {
+    if (p.length2() > mesh_.getSquaredMaxRadius()){
+        return false;
+    }
+
+    int intersect_count = 0;
+
+    // determine plucker coordinates of line p
+    Vector3 p_out = Vector3(1.1 * mesh_.getMaxRadius(), 0, 0);
+    Vector3 p_U = p-p_out;
+    Vector3 p_V = p.cross(p_out);
+
+    // load triangles
+    const std::vector<geo::Vector3>& t_points = mesh_.getPoints();
+    const std::vector<TriangleI>& triangles_i = mesh_.getTriangleIs();
+    for(std::vector<TriangleI>::const_iterator it = triangles_i.begin(); it != triangles_i.end(); ++it) {
+        Vector3 v1 = t_points[it->i1_];
+        Vector3 v2 = t_points[it->i2_];
+        Vector3 v3 = t_points[it->i3_];
+
+        Vector3 e1_U = v1-v2;
+        Vector3 e2_U = v2-v3;
+        Vector3 e3_U = v3-v1;
+
+        Vector3 e1_V = v1.cross(v2);
+        Vector3 e2_V = v2.cross(v3);
+        Vector3 e3_V = v3.cross(v1);
+
+        double S1 = side_operator(p_U, p_V, e1_U, e1_V);
+        double S2 = side_operator(p_U, p_V, e2_U, e2_V);
+        double S3 = side_operator(p_U, p_V, e3_U, e3_V);
+
+        if ((S1<0 && S2<0 && S3<0) || (S1>0 && S2>0 && S3>0)){ // the line passes through the triangle. now check the line segment
+            Vector3 l1_U = p_out-v1;
+            Vector3 l2_U = v1 - p;
+
+            Vector3 l1_V = p_out.cross(v1);
+            Vector3 l2_V = v1.cross(p);
+
+            double S4 = side_operator(l1_U, l1_V, e2_U, e2_V);
+            double S5 = side_operator(l2_U, l2_V, e2_U, e2_V);
+
+            if ((S4>0 && S5<0) || (S4<0 && S5>0)){
+                if (S1<0 && S2<0 && S3<0){
+                    intersect_count--;
+                }
+                else if(S1>0 && S2>0 && S3>0){
+                    intersect_count++;
+                }
+            }
+        }
+    }
+
+    if (intersect_count<0){
+        std::cout << "intersect_count lower than 0! Something went wrong!" << std::endl;
+        return false;
+    }
```

Comment:
This return statement has the same behaviour as the final return, so this could be removed. Also there needs to be a space between the if condition and the curly bracket. Also in the if condition there need to be spaces.

Reply:
True however this is to check our result if this condition holds it should record something in the log because something went wrong. And since this is only executed once in the function these extra operations wont slow it down too much